codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/RetryConfigDecorator.kt

```diff
@@ -0,0 +1,158 @@
+
+ */
+
+class RetryConfigDecorator : RustCodegenDecorator {
+    override val name: String = "RetryConfig"
+    override val order: Byte = 0
+
+    override fun configCustomizations(
+        codegenContext: CodegenContext,
+        baseCustomizations: List<ConfigCustomization>
+    ): List<ConfigCustomization> {
+        return baseCustomizations + RetryConfigProviderConfig(codegenContext.runtimeConfig)
+    }
+
+    override fun operationCustomizations(
+        codegenContext: CodegenContext,
+        operation: OperationShape,
+        baseCustomizations: List<OperationCustomization>
+    ): List<OperationCustomization> {
+        return baseCustomizations + RetryConfigConfigPlugin()
+    }
+
+    override fun libRsCustomizations(
+        codegenContext: CodegenContext,
+        baseCustomizations: List<LibRsCustomization>
+    ): List<LibRsCustomization> {
+        return baseCustomizations + PubUseRetryConfig(codegenContext.runtimeConfig)
+    }
+}
+
+class RetryConfigProviderConfig(runtimeConfig: RuntimeConfig) : ConfigCustomization() {
+    private val retryConfig = smithyTypesRetry(runtimeConfig)
+    private val codegenScope = arrayOf("RetryConfig" to retryConfig.member("RetryConfig"))
+    override fun section(section: ServiceConfig) = writable {
+        when (section) {
+            is ServiceConfig.ConfigStruct -> rustTemplate("pub(crate) retry_config: Option<#{RetryConfig}>,", *codegenScope)
+            is ServiceConfig.ConfigImpl -> emptySection
+            is ServiceConfig.BuilderStruct ->
+                rustTemplate("retry_config: Option<#{RetryConfig}>,", *codegenScope)
+            ServiceConfig.BuilderImpl ->
+                rustTemplate(
+                    """
+            pub fn retry_config(mut self, retry_config: #{RetryConfig}) -> Self {
+                self.set_retry_config(Some(retry_config));
+                self
+            }
+        
+            pub fn set_retry_config(&mut self, retry_config: Option<#{RetryConfig}>) -> &mut Self {
+                self.retry_config = retry_config;
+                self
+            }
+            """,
+                    *codegenScope
+                )
+            ServiceConfig.BuilderBuild -> rustTemplate(
+                """retry_config: self.retry_config,""",
+                *codegenScope
+            )
+        }
+    }
+}
+
+class RetryConfigConfigPlugin : OperationCustomization() {
+    override fun section(section: OperationSection): Writable {
+        return when (section) {
+            is OperationSection.MutateRequest -> writable {
+                // Allow the retry config to be late-inserted via another method
+                rust(
+                    """
+                if let Some(retry_config) = &${section.config}.retry_config {
+                    ${section.request}.properties_mut().insert(retry_config.clone());
+                }
+                """
+                )
+            }
+            else -> emptySection
+        }
+    }
+}
```

Comment:
since the retry config is not currently read out of the property bag, so we probably don't need this. I might imagine that at some point in the future we actually have the retry policy pull it out of the request but since that isn't supported now, probably best to remove this to avoid confusion

(Region is read out of the property bag to produce an endpoint and sign a request, for context)